python/cudf/cudf/core/cut.py

```diff
@@ -0,0 +1,277 @@
+    ...        index=['a', 'b', 'c', 'd', 'e'])
+    >>> pd.cut(s, 3)
+    """
+    left_inclusive = False
+    right_inclusive = True
+    # saving the original input x for use in case its a series
+    orig_x = x
+
+    if not ordered and labels is None:
+        raise ValueError("'labels' must be provided if 'ordered = False'")
+
+    if duplicates not in ["raise", "drop"]:
+        raise ValueError(
+            "invalid value for 'duplicates' parameter, valid options are: "
+            "raise, drop"
+        )
+
+    # bins can either be an int, sequence of scalars or an intervalIndex
+    if isinstance(bins, Sequence):
+        if len(set(bins)) is not len(bins):
+            if duplicates == "raise":
+                raise ValueError(
+                    f"Bin edges must be unique: {repr(bins)}.\n"
+                    f"You can drop duplicate edges by setting the 'duplicates'"
+                    "kwarg"
+                )
+            elif duplicates == "drop":
+                # get unique values but maintain list dtype
+                bins = cupy.unique(bins).tolist()
+
+    # if bins is an intervalIndex we ignore the value of right
+    if (
+        right is False
+        and isinstance(bins, pandas_IntervalIndex)
+        and bins.closed == "right"
+    ):
+        right = True
+
+    # create bins if given an int or single scalar
+    if not isinstance(bins, pandas_IntervalIndex):
+        if not isinstance(bins, (Sequence)):
+            if isinstance(
+                x, (pd.Series, cudf.Series, np.ndarray, cupy.ndarray)
+            ):
+                # changing to scalars and using sequence to get the bins
+                # because this allows masked arrays from value_counts to
+                # also be able to be handled by cut correctly
+                # pandas by default seems to turn all bins into a float
+                mn = cudf.Scalar(x.min(), dtype="float64")
+                mx = cudf.Scalar(x.max(), dtype="float64")
+            else:
+                mn = cudf.Scalar(min(x), dtype="float64")
+                mx = cudf.Scalar(max(x), dtype="float64")
+            step = cudf.Scalar((mx - mn) / bins, dtype="float64")
+            bins = sequence(
+                size=bins + 1, init=mn.device_value, step=step.device_value
+            ).values
+            adj = (mx - mn).value * 0.001
+            if right:
+                bins[0] -= adj
+            else:
+                bins[-1] += adj
+
+        # if right and include lowest we adjust the first
+        # bin edge to make sure it is included
+        if right and include_lowest:
+            bins[0] = bins[0] - 10 ** (-precision)
+
+        # adjust bin edges decimal precision
+        bins = cupy.around(bins, precision)
+
+        # if right is false the last bin edge is not included
+        if right is False:
+            right_edge = bins[len(bins) - 1]
+            x = cupy.asarray(x)
+            x[x == right_edge.item()] = right_edge.item() + 1
```

Comment:
Another argument of handling `bins` on host. Or at least we should cache `right_edge.item()` here.